src/main/java/ch/njol/skript/SkriptEventHandler.java

```diff
@@ -79,71 +80,68 @@ public PriorityListener(EventPriority priority) {
 	}
 
 	/**
-	 * A list tracking what Triggers are paired with what Events.
+	 * A Multimap tracking what Triggers are paired with what Events.
+	 * Each Event effectively maps to an ArrayList of Triggers.
 	 */
-	private static final List<NonNullPair<Class<? extends Event>, Trigger>> triggers = new ArrayList<>();
+	private static final Multimap<Class<? extends Event>, Trigger> triggers = ArrayListMultimap.create();
 
 	/**
-	 * A utility method to get all Triggers paired with the provided Event class.
+	 * A utility method to get all Triggers registered under the provided Event class.
 	 * @param event The event to find pairs from.
-	 * @return An iterator containing all Triggers paired with the provided Event class.
+	 * @return A List containing all Triggers registered under the provided Event class.
 	 */
-	private static Iterator<Trigger> getTriggers(Class<? extends Event> event) {
+	private static List<Trigger> getTriggers(Class<? extends Event> event) {
 		HandlerList eventHandlerList = getHandlerList(event);
 		assert eventHandlerList != null; // It had one at some point so this should remain true
-		return new ArrayList<>(triggers).stream()
-			.filter(pair -> pair.getFirst().isAssignableFrom(event) && eventHandlerList == getHandlerList(pair.getFirst()))
-			.map(NonNullPair::getSecond)
-			.iterator();
+		return triggers.asMap().entrySet().stream()
```

Comment:
```suggestion
		return triggers.entries().stream()
```

Reply:
That's exactly what I don't want. `entries()` is essentially a big collection of pairs, one to one. `asMap().entrySet()` is the events tied to their collections, one to many.

Reply:
You're doing the same thing but making a map every time. The old implementation had a paid object to get through.

Reply:
The point is I'm filtering before I'm getting all the triggers in the valid entries. If I were to get use `entries()`, I'd be filtering duplicate events, and we're back to the same bad structure that was in the original implementation.

Additionally, I wondered about the cost of `asMap()`, and it turns out it's O(1), so I don't think it's an issue. It was discussed in the skriptlang-development channel.